mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala

```diff
@@ -587,18 +722,69 @@ class SparseMatrix @Since("2.0.0") (
     }
   }
 
+  override def numNonzeros: Int = values.count(_ != 0)
+
+  override def numActives: Int = values.length
+
   /**
-   * Generate a `DenseMatrix` from the given `SparseMatrix`. The new matrix will have isTransposed
-   * set to false.
+   * Generate a `SparseMatrix` from this `SparseMatrix`, removing explicit zero values if they
+   * exist.
+   *
+   * @param colMajor Whether or not the resulting `SparseMatrix` values are in column major
+   *                    order.
    */
-  @Since("2.0.0")
-  def toDense: DenseMatrix = {
-    new DenseMatrix(numRows, numCols, toArray)
+  private[ml] override def toSparseMatrix(colMajor: Boolean): SparseMatrix = {
+    if (!(colMajor ^ isTransposed)) {
+      // breeze transpose rearranges values in column major and removes explicit zeros
+      if (!isTransposed) {
+        // it is row major and we want col major
+        val breezeTransposed = asBreeze.asInstanceOf[BSM[Double]].t
+        Matrices.fromBreeze(breezeTransposed).transpose.asInstanceOf[SparseMatrix]
+      } else {
+        // it is col major and we want row major
+        val breezeTransposed = asBreeze.asInstanceOf[BSM[Double]]
+        Matrices.fromBreeze(breezeTransposed).asInstanceOf[SparseMatrix]
+      }
+    } else {
```

Comment:
Can we document here that it's when the layout of this and colMajor is different? Easier read than `(colMajor ^ isTranspose)` condition here. Even more readable to use pattern matching with exact boolean on both variables.